### PR TITLE
Fix recreation of user in Che DB, when it has been recreated on Keycloak

### PR DESCRIPTION
**Pull request by mkuznyetsov**
### What does this PR do?
When user has been removed and created again with the same email, on it's log Che will make sure to recreate one properly in its DB. The detailed workflow can be seen here: https://github.com/eclipse/che/issues/8081#issuecomment-376502397

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/8081

### CODE REVIEW

**Comment by mkuznyetsov** [comment]
ci-test

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserManager.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+import org.eclipse.che.api.core.model.user.User;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.user.server.event.BeforeUserRemovedEvent;
+import org.eclipse.che.api.user.server.event.PostUserPersistedEvent;
+import org.eclipse.che.api.user.server.model.impl.UserImpl;
+import org.eclipse.che.api.user.server.spi.PreferenceDao;
+import org.eclipse.che.api.user.server.spi.ProfileDao;
+import org.eclipse.che.api.user.server.spi.UserDao;
+import org.eclipse.che.multiuser.api.account.personal.PersonalAccountUserManager;
+
```
Comment (skabashnyuk) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserManager.java:37]: javadoc

**Review by skabashnyuk — COMMENTED** [review]
Do you have CI results

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserManager.java (hunk 1 above)
Comment (mkuznyetsov) [reply multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserManager.java:37]: ok

**Comment by mkuznyetsov** [comment]
ci-test

**Review by skabashnyuk — APPROVED** [review]

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserManager.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+/**
+ * Extension of User Manager, providing utility operations related to Keycloak User management, and
+ * overriding create/remove operations to be compatible with {@link
+ * org.eclipse.che.multiuser.keycloak.server.dao.KeycloakProfileDao}
+ *
+ * @author Mykhailo Kuznietsov
+ */
+public class KeycloakUserManager extends PersonalAccountUserManager {
```
Comment (akorneta) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserManager.java]: consider add `@Singleton` annotation or bind `KeycloakUserManager` in singleton scope

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserManager.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+   * way, there will be no conflict with existing user id or email upon recreation. In case of
+   * conflict with user name, it may be prepended randomized symbols
+   *
+   * @param id - user id from JWT token
+   * @param email - user email from JWT token
+   * @param username - user name from JWT token
+   * @return user object from Che Database, with all needed actualization operations performed on
+   *     him
+   */
+  public User getOrCreateUser(String id, String email, String username)
```
Comment (akorneta) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserManager.java]: add documentation when `ServerException` and `ConflictException` may occur

**Review by mshaposhnik — APPROVED** [review]
Ok for me with previous comments fix.

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserManager.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+      return;
+    }
+    preferencesDao.remove(id);
+    eventService.publish(new BeforeUserRemovedEvent(user)).propagateException();
+    userDao.remove(id);
+  }
+
+  /**
+   * Method is used to retrieve user object from Che DB for given user {@code id}, {@code email},
+   * and {@code username}, that would appear in Keycloak JWT token. Various actualization operations
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserManager.java]: I think `KeycloakUserManager` knows nothing about `JWT` tokens. Please consider fixing docs.

**Review by sleshchenko — APPROVED** [review]
LGTM
Please take a look comments

**Review by akorneta — APPROVED** [review]

**Pull request merged**